### PR TITLE
Revamp of Dataset and YAXArray saving

### PR DESCRIPTION
**Pull request by meggart**
New features include:

- `savedataset` function
- `savecube` now calls `savedataset` after transforming with `to_dataset`
- `append` option for savedataset to add variables to an existing store
- optimized writing, so this can be used for rechunking data
- lots of examples in the docs

In addition, I implemented something I had planned a long time ago: explicitly adding the chunks of a YAXArray as a field and an option for users to modify the chunking using `setchunks`. This way, to store a dataset with user-defined chunking, one just calls `setchunks` prior to saving the dataset. Another area of application is when `map`, `concatenatecubes`, `mapCube` or `CubeTable` and friends fail to find a good common chunking when operating on multiple cubes. Then the user has always the possiblity to reset the chunks that YAXArray sees and thereby give hints on how to best access the data.

### CODE REVIEW

**Comment by meggart** [comment]
@felixcremer it would be great if you could experiment a bit, in particular if this is useful for rechunking, or if you encounter problems.

File: src/Cubes/Cubes.jl [hunk 1]
```diff
@@ -81,8 +81,9 @@ struct YAXArray{T,N,A<:AbstractArray{T,N},AT}
     axes::AT
     data::A
     properties::Dict{String}
+    chunks::GridChunks{N}
     cleaner::Vector{CleanMe}
-    function YAXArray(axes, data, properties, cleaner)
+    function YAXArray(axes, data, properties, chunks, cleaner)
```
Comment (felixcremer) [inline src/Cubes/Cubes.jl:86]: Isn't this change breaking? Because I can't do `YAXArray(axes, data, properties ,cleaner)` anymore.

**Comment by felixcremer** [comment]
This is a cube that could led to the error above:
```julia
julia> a = YAXArray([RangeAxis(:Lon, 1:400), RangeAxis(:Lat, 1:300), CategoricalAxis("Polarisation", 1:2), CategoricalAxis("IMF", 1:8), CategoricalAxis("Variable", string.(1:4))], rand(400,300,2, 8,4))

```

We had different errors depending on the name of the "Variable" dimensions and also depending on the length of the strings in the Axis values.

**Comment by felixcremer** [comment]
I think, I found the error for the not aligned chunks and the two cubes that I want to concatenate have the same chunksizes, but have different offsets:
```julia
julia> DiskArrays.eachchunk(setchunks(permutedims(haidescsub, (3,1,2,4)), chunks))[1,1,:,1]
15-element Vector{NTuple{4, UnitRange{Int64}}}:
 (1:358, 1:401, 1:20, 1:1)
 (1:358, 1:401, 21:40, 1:1)
 (1:358, 1:401, 41:60, 1:1)
 (1:358, 1:401, 61:80, 1:1)
 (1:358, 1:401, 81:100, 1:1)
 (1:358, 1:401, 101:120, 1:1)
 (1:358, 1:401, 121:140, 1:1)
 (1:358, 1:401, 141:160, 1:1)
 (1:358, 1:401, 161:180, 1:1)
 (1:358, 1:401, 181:200, 1:1)
 (1:358, 1:401, 201:220, 1:1)
 (1:358, 1:401, 221:240, 1:1)
 (1:358, 1:401, 241:260, 1:1)
 (1:358, 1:401, 261:280, 1:1)
 (1:358, 1:401, 281:299, 1:1)

julia> DiskArrays.eachchunk(su)[1,1,:,1]
sub1       subset      subset!     subsetcube  subtypes    success     sum         sum!        summary     supertype   supertypes  surface     surface!
julia> DiskArrays.eachchunk(sub1)[1,1,:,1]
16-element Vector{NTuple{4, UnitRange{Int64}}}:
 (1:358, 1:401, 1:10, 1:1)
 (1:358, 1:401, 11:30, 1:1)
 (1:358, 1:401, 31:50, 1:1)
 (1:358, 1:401, 51:70, 1:1)
 (1:358, 1:401, 71:90, 1:1)
 (1:358, 1:401, 91:110, 1:1)
 (1:358, 1:401, 111:130, 1:1)
 (1:358, 1:401, 131:150, 1:1)
 (1:358, 1:401, 151:170, 1:1)
 (1:358, 1:401, 171:190, 1:1)
 (1:358, 1:401, 191:210, 1:1)
 (1:358, 1:401, 211:230, 1:1)
 (1:358, 1:401, 231:250, 1:1)
 (1:358, 1:401, 251:270, 1:1)
 (1:358, 1:401, 271:290, 1:1)
 (1:358, 1:401, 291:299, 1:1)
``

**Comment by felixcremer** [comment]
Unfortunately I have no idea how to move forward from here.
For posterity, this is the error that I get when I try to concatenate two cubes and I already use setchunks to the same chunks as the first cube:

```julia
julia> concatenatecubes([sub1, setchunks(permutedims(haidescsub, (3,1,2,4)), chunks)], catax)
(i, eachchunk(cl[i])) = (2, [(1:358, 1:401, 1:20, 1:1);;; (1:358, 1:401, 21:40, 1:1);;; (1:358, 1:401, 41:60, 1:1);;; (1:358, 1:401, 61:80, 1:1);;; (1:358, 1:401, 81:100, 1:1);;; (1:358, 1:401, 101:120, 1:1);;; (1:358, 1:401, 121:140, 1:1);;; (1:358, 1:401, 141:160, 1:1);;; (1:358, 1:401, 161:180, 1:1);;; (1:358, 1:401, 181:200, 1:1);;; (1:358, 1:401, 201:220, 1:1);;; (1:358, 1:401, 221:240, 1:1);;; (1:358, 1:401, 241:260, 1:1);;; (1:358, 1:401, 261:280, 1:1);;; (1:358, 1:401, 281:299, 1:1);;;; (1:358, 1:401, 1:20, 2:2);;; (1:358, 1:401, 21:40, 2:2);;; (1:358, 1:401, 41:60, 2:2);;; (1:358, 1:401, 61:80, 2:2);;; (1:358, 1:401, 81:100, 2:2);;; (1:358, 1:401, 101:120, 2:2);;; (1:358, 1:401, 121:140, 2:2);;; (1:358, 1:401, 141:160, 2:2);;; (1:358, 1:401, 161:180, 2:2);;; (1:358, 1:401, 181:200, 2:2);;; (1:358, 1:401, 201:220, 2:2);;; (1:358, 1:401, 221:240, 2:2);;; (1:358, 1:401, 241:260, 2:2);;; (1:358, 1:401, 261:280, 2:2);;; (1:358, 1:401, 281:299, 2:2)])
(i, chunks) = (2, [(1:358, 1:401, 1:10, 1:1);;; (1:358, 1:401, 11:30, 1:1);;; (1:358, 1:401, 31:50, 1:1);;; (1:358, 1:401, 51:70, 1:1);;; (1:358, 1:401, 71:90, 1:1);;; (1:358, 1:401, 91:110, 1:1);;; (1:358, 1:401, 111:130, 1:1);;; (1:358, 1:401, 131:150, 1:1);;; (1:358, 1:401, 151:170, 1:1);;; (1:358, 1:401, 171:190, 1:1);;; (1:358, 1:401, 191:210, 1:1);;; (1:358, 1:401, 211:230, 1:1);;; (1:358, 1:401, 231:250, 1:1);;; (1:358, 1:401, 251:270, 1:1);;; (1:358, 1:401, 271:290, 1:1);;; (1:358, 1:401, 291:299, 1:1);;;; (1:358, 1:401, 1:10, 2:2);;; (1:358, 1:401, 11:30, 2:2);;; (1:358, 1:401, 31:50, 2:2);;; (1:358, 1:401, 51:70, 2:2);;; (1:358, 1:401, 71:90, 2:2);;; (1:358, 1:401, 91:110, 2:2);;; (1:358, 1:401, 111:130, 2:2);;; (1:358, 1:401, 131:150, 2:2);;; (1:358, 1:401, 151:170, 2:2);;; (1:358, 1:401, 171:190, 2:2);;; (1:358, 1:401, 191:210, 2:2);;; (1:358, 1:401, 211:230, 2:2);;; (1:358, 1:401, 231:250, 2:2);;; (1:358, 1:401, 251:270, 2:2);;; (1:358, 1:401, 271:290, 2:2);;; (1:358, 1:401, 291:299, 2:2)])
size(chunks) = (1, 1, 16, 2)
size(eachchunk(cl[i])) = (1, 1, 15, 2)
ERROR: Trying to concatenate cubes with different chunk sizes. Consider manually setting a common chunk size using `setchunks`.
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:33
 [2] concatenatecubes(cl::Vector{YAXArray{Union{Missing, Float32}, 4, A, Vector{CubeAxis}} where A<:AbstractArray{Union{Missing, Float32}, 4}}, cataxis::CategoricalAxis{String, :IMF, Vector{String}})
   @ YAXArrays.Cubes ~/Documents/papers_wip/EMDAmazonas/dev/YAXArrays/src/Cubes/TransformedCubes.jl:48
 [3] top-level scope
   @ REPL[103]:1
``

**Comment by felixcremer** [comment]
It doesn't find the backend from the filename. Maybe this only happens for .zarr files.

**Comment by lazarusA** [comment]
Saving a dataset and then reopening it leads to the following error:
```Julia
julia> ds2 = open_dataset("data/concatedds.zarr")
ERROR: MethodError: no method matching typemax(::Type{Union{Missing, Float32}})
Closest candidates are:
  typemax(::Union{DateTime, Type{DateTime}}) at /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Dates/src/types.jl:453
  typemax(::Union{Date, Type{Date}}) at /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Dates/src/types.jl:455
  typemax(::Union{Dates.Time, Type{Dates.Time}}) at /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/Dates/src/types.jl:457
  ...
Stacktrace:
 [1] DiskArrayTools.CFDiskArray(a::ZArray{Union{Missing, Float32}, 3, Zarr.BloscCompressor, DirectoryStore}, attr::Dict{String, Any})
   @ DiskArrayTools ~/.julia/packages/DiskArrayTools/rEAgw/src/DiskArrayTools.jl:164
 [2] open_dataset(g::String; driver::Symbol)
   @ YAXArrays.Datasets ~/Documents/SeasFire/dev/YAXArrays/src/DatasetAPI/Datasets.jl:272
 [3] open_dataset(g::String)
   @ YAXArrays.Datasets ~/Documents/SeasFire/dev/YAXArrays/src/DatasetAPI/Datasets.jl:242
 [4] top-level scope
   @ REPL[2]:1
```

**Comment by meggart** [comment]
@lazarusA are you on this PR branch or do you get the error on master?

**Comment by meggart** [comment]
@felixcremer could you please push the changes to update the PR?

**Comment by lazarusA** [comment]
@meggart  yes, this is on this branch. Right @felixcremer? 
We were trying to save a new cube with attributes attached to new variables. Like `units`, `long_name`, attributes that you normally have in .nc files without success.

**Comment by felixcremer** [comment]
Yes this was on this branch. I also get this from the Gdalcube we optimzed yesterday. You can open the cube, when you use nonmissingtype in the DiskarrayTools function that throws the error. But I am not sure, whether this is the way
 to fix this.

**Comment by meggart** [comment]
Can you please send me a full path to a file you are trying to open?

**Comment by felixcremer** [comment]
The error happened with a local fluxcom dataset and after saving it with this branch.

**Comment by meggart** [comment]
I meant an MWE, it is really hard to reproduce without

**Comment by felixcremer** [comment]
This is superseeded by #150

**Pull request closed without merging**